### PR TITLE
chore: Updating provider factories with new market map configuration

### PR DESCRIPTION
**Pull request by davidterpay**

### CODE REVIEW

File: cmd/oracle/main.go [hunk 1]
```diff
@@ -82,19 +83,11 @@ func main() {
...
 		return
 	}
-
-	// Create the conversion market aggregator.
-	aggregator, err := oraclemath.NewMedianAggregator(logger, cfg.Market)
-	if err != nil {
-		logger.Error("failed to create median aggregator", zap.Error(err))
-		return
-	}
-
```
Comment (davidterpay) [inline cmd/oracle/main.go:92]: I am going to swap this out till the refactor with the conversion logic done.

File: cmd/oracle/main.go [hunk 2]
```diff
@@ -109,9 +102,7 @@ func main() {
 	// cancel oracle on interrupt or terminate
 	go func() {
 		<-sigs
-		logger.Info(
-			"received interrupt or terminate signal, closing oracle",
-		)
+		logger.Info("received interrupt or terminate signal, closing oracle")
```
Comment (aljo242) [inline cmd/oracle/main.go]:
```suggestion
		logger.Info("received interrupt or terminate signal; closing oracle")
```

File: config/local/README.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 ## Overview
 
-This directory contains the [configuration file](./oracle.toml) for the local oracle instance. To update the configuration, edit the `LocalConfig` in [generate_toml.go](./generate_toml.go) and run the following command:
+This directory contains the [configuration file](./oracle.json) for the local oracle instance as well as [market map file](./market.json) that contains all of the markets that are supported by the oracle. To update the set of provider's utilized, update the `LocalOracleConfig` in `generate.go`. To update the set of markets supported, update the `ProvidersToMarkets` in `generate.go`. To add a custom conversion path, update the `TickerPaths` in `generate.go`.
```
Comment (aljo242) [inline config/local/README.md]:
```suggestion
This directory contains the [configuration file](./oracle.json) for the local oracle instance as well as [market map file](./market.json) that contains all markets that are supported by the oracle. To update the set of provider's utilized, update the `LocalOracleConfig` in `generate.go`. To update the set of markets supported, update the `ProvidersToMarkets` in `generate.go`. To add a custom conversion path, update the `TickerPaths` in `generate.go`.
```

File: config/local/generate.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	if err := encoder.Encode(marketMap); err != nil {
+		panic(err)
+	}
+}
+
+// createMarketMap creates a market map given all of the local market configurations for
+// each provider as well as the custom conversion markets. We do so to ensure that the
+// oracle is always started using the market map that is expected to be stored by the
+// market map module.
+func createMarketMap() (mmtypes.MarketMap, error) {
```
Comment (aljo242) [inline config/local/generate.go:230]: could we use this kind of function to generate a large `genesis` field for testing?

File: oracle/types/market.go [hunk 5]
```diff
@@ -62,6 +92,15 @@ func NewProviderMarketMap(name string, tickerConfigs TickerToProviderConfig) (Pr
 	}, nil
 }
 
+// GetTickers returns the tickers from the provider market map.
+func (pmm ProviderMarketMap) GetTickers() []mmtypes.Ticker {
```
Comment (aljo242) [inline oracle/types/market.go]:
```suggestion
func (pmm *ProviderMarketMap) GetTickers() []mmtypes.Ticker {
```

File: oracle/types/market.go [hunk 6]
```diff
@@ -62,6 +92,15 @@ func NewProviderMarketMap(name string, tickerConfigs TickerToProviderConfig) (Pr
...
+	tickers := make([]mmtypes.Ticker, 0, len(pmm.TickerConfigs))
+	for ticker := range pmm.TickerConfigs {
+		tickers = append(tickers, ticker)
+	}
+	return tickers
+}
+
 // ValidateBasic performs basic validation on the provider market map.
 func (pmm ProviderMarketMap) ValidateBasic() error {
```
Comment (aljo242) [inline oracle/types/market.go]:
```suggestion
func (pmm *ProviderMarketMap) ValidateBasic() error {
```

File: oracle/types/market.go [hunk 7]
```diff
@@ -98,3 +136,28 @@ func (pmm ProviderMarketMap) ValidateBasic() error {
...
+	if err != nil {
+		return config, fmt.Errorf("error reading config file: %w", err)
+	}
+
+	// Unmarshal the JSON data into the config struct
+	if err := json.Unmarshal(data, &config); err != nil {
+		return config, fmt.Errorf("error unmarshalling config JSON: %w", err)
+	}
+
+	// Optionally, validate the config if needed
```
Comment (aljo242) [inline oracle/types/market.go]:
```suggestion
```

File: oracle/types/market.go [hunk 8]
```diff
@@ -98,3 +136,28 @@ func (pmm ProviderMarketMap) ValidateBasic() error {
...
+	if err := config.ValidateBasic(); err != nil {
+		return config, fmt.Errorf("error validating config: %w", err)
+	}
+
+	// Return the populated config struct
```
Comment (aljo242) [inline oracle/types/market.go]:
```suggestion
```

File: providers/factories/oracle/websocket.go [hunk 9]
```diff
@@ -30,93 +28,96 @@ import (
...
-			wsDataHandler  wshandlers.WebSocketDataHandler[slinkytypes.CurrencyPair, *big.Int]
-			connHandler    wshandlers.WebSocketConnHandler
-		)
+func WebSocketQueryHandlerFactory(
+	logger *zap.Logger,
+	cfg config.ProviderConfig,
+	wsMetrics wsmetrics.WebSocketMetrics,
+	marketMap types.ProviderMarketMap,
+) (types.PriceWebSocketQueryHandler, error) {
+	// Validate the provider config.
```
Comment (aljo242) [inline providers/factories/oracle/websocket.go]:
```suggestion
```

File: x/marketmap/types/market.go [hunk 10]
```diff
@@ -1,25 +1,32 @@
...
+// 1. Each ticker has a corresponding provider supporting it.
+// 2. Each ticker is valid.
+// 3. Each provider is valid.
+// 4. Each path is valid.
+// 5. Each operation (ticker) in each path is supported by the market map.
 func (mm *MarketMap) ValidateBasic() error {
-	if len(mm.Paths) != len(mm.Tickers) || len(mm.Paths) != len(mm.Providers) {
-		return fmt.Errorf("length of Tickers, Paths and Providers must be equal")
+	if len(mm.Tickers) != len(mm.Providers) {
+		return fmt.Errorf("each ticker must have a corresponding provider supporting it")
```
Comment (aljo242) [inline x/marketmap/types/market.go]:
```suggestion
		return fmt.Errorf("each ticker must have a corresponding provider list supporting it")
```

**Review by aljo242 — APPROVED** [review]
lgtm with some tiny nits

**Pull request merged**